File: Core/Queries/DTOs/EntityTableDto.cs
namespace FinanceTracker.Core.Queries.DTOs;

public record EntityTableDto(
    IReadOnlyCollection<EntityColumnDto> Columns,
    IReadOnlyCollection<ValueHistoryRecordDto> Rows
);

public record EntityColumnDto(
    Guid Id,
    string Name,
    string? ParentName,
    Guid? DefaultPhysicalAllocationId
);

public record ValueHistoryRecordDto(
    string Key,
    IReadOnlyCollection<EntityValueSnapshotDto?> Entities,
    ValueSnapshotDto Summary,
    // Extras
    TargetDto? Target = null,
    ScoreDto? Score = null
);

public record TargetDto(
    decimal TargetInMainCurrency,
    decimal Percentage
);

public record ScoreDto(
    decimal ChangePercent,
    InflationDto? Inflation,
    decimal TotalChangePercent
);

public record InflationDto(
    decimal Value,
    bool Confirmed
);

File: Web/ClientApp/src/api/value-history/DTOs/EntityTableDto.ts
import {EntityValueSnapshotDto, ValueSnapshotDto} from "./ValueSnapshotDto";

export type EntityTableDto = {
    columns: EntityColumnDto[],
    rows: ValueHistoryRecordDto[]
}

export type EntityColumnDto = {
    id: string,
    name: string,
    parentName: string | undefined,
    defaultPhysicalAllocationId: string | undefined
}

export type ValueHistoryRecordDto = {
    key: string,
    entities: (EntityValueSnapshotDto | undefined)[],
    summary: ValueSnapshotDto | undefined
    target: TargetDto | undefined,
    score: ScoreDto | undefined,
    // Used internally
    newEntry?: boolean,
}

export type TargetDto = {
    targetInMainCurrency: number,
    percentage: number,
}

export type ScoreDto = {
    changePercent: number,
    inflation: InflationDto | undefined,
    totalChangePercent: number
}

export type InflationDto = {
    value: number,
    confirmed: boolean
}